Source/WebKit/UIProcess/_WKWebViewPrintFormatter.mm

```diff
@@ -66,6 +71,33 @@ - (WKWebView *)_webView
     return static_cast<WKWebView *>(view);
 }
 
+- (CGPDFDocumentRef)_printedDocument
+{
+    if (self.requiresMainThread)
```

Comment:
It's a bit odd that `_WKWebViewPrintFormatter` exclusively checks for `requiresMainThread` while `WKWebView` only checks for `isMainRunLoop`.  Should we be checking `isMainRunLoop` here too in case `requiresMainThread` is `NO` but the `_WKWebViewPrintFormatter` is called on the main thread anyways?

Reply:
No, we want to grab the lock, since PDF callback (writing the printed document) will always be called on the main thread, but the drawing can happen on a background thread (reading the printed document) .

We could optimize away the lock if the print formatter is used on the main thread only (the `UIPrintPageRenderer` case, which we can tell by where `_recalcPageCount` is called). However, since this is not a performance critical path, and the optimization would make the logic slightly more complex, I have kept the check simple. Printing a loaded PDF will avoid locking though.